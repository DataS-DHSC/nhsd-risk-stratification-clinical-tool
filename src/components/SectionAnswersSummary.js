import React from 'react';
import PropTypes from 'prop-types';
import { SummaryList, Button } from 'nhsuk-react-components';
import { useRouter } from 'next/router';
import { addNamespace } from '../utils/namespacingTools';

const SectionAnswersSummary = ({ pageSection, data, pageUrl }) => {
  const fields = Object.entries(pageSection.fields).filter(
    ([fieldId]) =>
      (data[addNamespace(fieldId)] !== pageSection.fields[fieldId].default ||
        pageSection.fields[fieldId].default === null) &&
      pageSection.fields[fieldId].type !== 'bmi-calc'
  );

  const router = useRouter();
  return (
    <SummaryList>
      {fields.map(([fieldId, field]) => {
        const {
          label: fieldLabel,
          type,
          options: fieldOptions,
          ariaLabel,
          fieldValidator,
        } = field;
        const fieldValue = data[addNamespace(fieldId)];

        const validationsMessage = fieldValidator
          ? fieldValidator(fieldValue)
          : '';

        let displayedValue = '';
        let selectedOption;

        if (fieldValue) {
          switch (type) {
            case 'multiradio':
              selectedOption = fieldOptions.find(
                (fieldOption) => fieldOption.name.toString() === fieldValue
              );
              displayedValue = selectedOption.label;
              break;
            case 'select':
              const selectFieldFlatOptions = fieldOptions.flatMap(
                (option) => option.options
              );
              selectedOption = selectFieldFlatOptions.find(
                (fieldOption) => fieldOption.name.toString() === fieldValue
              );
              displayedValue = selectedOption.label;

              break;
            case 'radio':
              displayedValue = fieldValue === 'true' ? 'Yes' : 'No';
              break;
            default:
              displayedValue = fieldValue;
          }
        }
        if (validationsMessage) {
          displayedValue = `${displayedValue} (${validationsMessage})`;
        }

        return (
          <SummaryList.Row key={fieldId} id={`summary-row-${fieldId}`}>
            <SummaryList.Key>{fieldLabel}</SummaryList.Key>
            <SummaryList.Value
              className={validationsMessage && 'nhsuk-validation-error'}
            >
              {displayedValue}
            </SummaryList.Value>
            <SummaryList.Actions>
              <Button
                type="submit"
                className="change-button"
                id={`change-button-${fieldId}`}
                formAction={pageUrl}
                aria-label={`Change ${
                  ariaLabel || fieldLabel
                } from ${displayedValue}`}
                onClick={(e) => {
                  e.preventDefault();
                  router.push(pageUrl);
                }}
              >
                Change
              </Button>
            </SummaryList.Actions>
          </SummaryList.Row>
        );
      })}
    </SummaryList>
  );
};

SectionAnswersSummary.propTypes = {
  pageSection: PropTypes.shape().isRequired,
  data: PropTypes.shape().isRequired,
  pageUrl: PropTypes.string.isRequired,
};

export default SectionAnswersSummary;
