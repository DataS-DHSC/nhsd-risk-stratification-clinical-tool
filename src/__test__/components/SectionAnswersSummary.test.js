import React from 'react';
import { render, fireEvent, createEvent } from '@testing-library/react';
import { useRouter } from 'next/router';
import SectionAnswersSummary from '../../components/SectionAnswersSummary';

jest.mock('next/router');

const setup = ({ fields, data, defaults }) => {
  const props = {
    formKey: 'test-form',
    pageSection: { fields },
    data,
    pageUrl: 'test_url',
    defaults,
  };

  return render(<SectionAnswersSummary {...props} />);
};

describe('SectionAnswersSummary', () => {
  afterEach(jest.clearAllMocks);

  it('should match snapshot if data is from multiradio', () => {
    const fields = {
      testFieldMultiRadio: {
        label: 'multiradio label',
        type: 'multiradio',
        options: [
          { name: 'option1Radio', label: 'option1RadioLabel' },
          { name: 'option2Radio', label: 'option2RadioLabel' },
        ],
      },
    };

    const container = setup({
      fields,
      data: { 'answers.testFieldMultiRadio': 'option2Radio' },
      defaults: {
        'test-form': {
          testFieldMultiRadio: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot if data is from select input', () => {
    const fields = {
      testFieldSelect: {
        label: 'select label',
        type: 'select',
        options: [
          {
            label: 'subgroup1',
            options: [
              { name: 'option11Select', label: 'option11Select' },
              { name: 'option12Select', label: 'option12Select' },
            ],
          },
          {
            label: 'subgroup2',
            options: [{ name: 'option21Select', label: 'option21Select' }],
          },
        ],
      },
    };

    const container = setup({
      fields,
      data: { 'answers.testFieldSelect': 'option21Select' },
      defaults: {
        'test-form': {
          testFieldSelect: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot if data is from radio input', () => {
    const fields = {
      testFieldRadioYes: {
        label: 'radio label yes',
        type: 'radio',
      },
      testFieldRadioNo: {
        label: 'radio label no',
        type: 'radio',
      },
    };

    const container = setup({
      fields,
      data: {
        'answers.testFieldRadioYes': 'true',
        'answers.testFieldRadioNo': 'false',
      },
      defaults: {
        'test-form': {
          testFieldRadio: null,
          testFieldRadioNo: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot if data is from text input', () => {
    const fields = {
      testFieldText: {
        label: 'label text',
        type: 'text-input',
      },
    };

    const container = setup({
      fields,
      data: { 'answers.testFieldText': 'testValue' },
      defaults: {
        'test-form': {
          testFieldText: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when validation fails', () => {
    const fields = {
      testFieldText: {
        label: 'label text',
        type: 'text-input',
        fieldValidator: () => 'validation message',
      },
    };

    const container = setup({
      fields,
      data: { 'answers.testFieldText': 'testValue' },
      defaults: {
        'test-form': {
          testFieldText: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when value equals default', () => {
    const fields = {
      testFieldText: {
        label: 'label text',
        type: 'text-input',
      },
    };

    const container = setup({
      fields,
      data: { 'answers.testFieldText': null },
      defaults: {
        'test-form': {
          testFieldText: null,
        },
      },
    });

    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should pass the correct url to the router when client side routing is enabled', () => {
    const router = { push: jest.fn() };
    useRouter.mockImplementationOnce(() => router);

    const fields = {
      testFieldRadio: {
        label: 'radio label',
        type: 'radio',
      },
    };

    const { container } = setup({
      fields,
      data: { 'answers.testFieldRadio': 'true' },
      clientSideRoutingEnabled: true,
      defaults: {
        'test-form': {
          testFieldRadio: null,
        },
      },
    });

    const element = container.querySelector('#change-button-testFieldRadio');
    const clickEvent = createEvent.click(element);
    fireEvent(element, clickEvent);

    expect(router.push).toHaveBeenCalledWith('test_url');
    expect(clickEvent.defaultPrevented).toBe(true);
  });
});
