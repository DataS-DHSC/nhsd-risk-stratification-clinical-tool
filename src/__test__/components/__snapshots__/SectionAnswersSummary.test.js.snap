// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionAnswersSummary should match snapshot if data is from multiradio 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldMultiRadio"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        multiradio label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        option2RadioLabel
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change multiradio label from option2RadioLabel"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldMultiRadio"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`SectionAnswersSummary should match snapshot if data is from radio input 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldRadioYes"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        radio label yes
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        Yes
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change radio label yes from Yes"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldRadioYes"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldRadioNo"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        radio label no
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        No
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change radio label no from No"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldRadioNo"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`SectionAnswersSummary should match snapshot if data is from select input 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldSelect"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        select label
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        option21Select
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change select label from option21Select"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldSelect"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`SectionAnswersSummary should match snapshot if data is from text input 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldText"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        label text
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      >
        testValue
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change label text from testValue"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldText"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`SectionAnswersSummary should match snapshot when validation fails 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldText"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        label text
      </dt>
      <dd
        class="nhsuk-summary-list__value nhsuk-validation-error"
      >
        testValue (validation message)
      </dd>
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change label text from testValue (validation message)"
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldText"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;

exports[`SectionAnswersSummary should match snapshot when value equals default 1`] = `
<DocumentFragment>
  <dl
    class="nhsuk-summary-list"
  >
    <div
      class="nhsuk-summary-list__row"
      id="summary-row-testFieldText"
    >
      <dt
        class="nhsuk-summary-list__key"
      >
        label text
      </dt>
      <dd
        class="nhsuk-summary-list__value"
      />
      <dd
        class="nhsuk-summary-list__actions"
      >
        <button
          aria-disabled="false"
          aria-label="Change label text from "
          class="nhsuk-button change-button"
          formaction="test_url"
          id="change-button-testFieldText"
          type="submit"
        >
          Change
        </button>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;
